src/tools/cephfs_mirror/PeerReplayer.cc

```diff
@@ -1319,6 +1321,150 @@ int PeerReplayer::do_synchronize(const std::string &dir_root, const Snapshot &cu
+                                 const boost::optional<Snapshot>& prev) {
+  if (!prev) {
+    derr << ": invalid previous snapshot" << dendl;
+    return -ENODATA;
+  }
+
+  dout(20) << ": incremental sync check from prev=" << prev << dendl;
+
+  FHandles fh;
+  int r = pre_sync_check_and_open_handles(dir_root, current, prev, &fh);
+  if (r < 0) {
+    dout(5) << ": cannot proceeed with sync: " << cpp_strerror(r) << dendl;
+    return r;
+  }
+
+  BOOST_SCOPE_EXIT_ALL( (this)(&fh) ) {
+    post_sync_close_handles(fh);
+  };
+
+  // record that we are going to "dirty" the data under this directory root
+  auto snap_id_str{stringify(current.second)};
+  r = ceph_setxattr(m_remote_mount, dir_root.c_str(), "ceph.mirror.dirty_snap_id",
+                    snap_id_str.c_str(), snap_id_str.size(), 0);
+  if (r < 0) {
+    derr << ": error setting \"ceph.mirror.dirty_snap_id\" on dir_root=" << dir_root
+         << ": " << cpp_strerror(r) << dendl;
+    return r;
+  }
+
+  ceph_snapdiff_info sd_info;
+  ceph_snapdiff_entry_t sd_entry;
+  std::string epath = ".";
+  std::string npath = "", nabs_path = "", nname = "";
+  int ntype = 0;
+
+  //the queue of SyncDiffEntry items (directories) to be synchronized
+  std::queue<SyncDiffEntry> sync_queue;
+
+  //start with initial/default entry
+  strncpy(sd_entry.dir_entry.d_name, ".", sizeof(sd_entry.dir_entry.d_name));
+  sd_entry.dir_entry.d_type = 4;
+  sync_queue.emplace(SyncDiffEntry(sd_entry, epath));
+
+  while (!sync_queue.empty()) {
+    if (should_backoff(dir_root, &r)) {
+      dout(0) << ": backing off r=" << r << dendl;
+      break;
+    }
+
+    dout(20) << ": " << sync_queue.size() << " entries in stack" << dendl;
+    const auto &queue_entry = sync_queue.front();
+    epath = queue_entry.path;
+    dout(20) << ": syncing entry, path=" << epath << dendl;
+    r = ceph_open_snapdiff(m_local_mount, dir_root.c_str(), epath.c_str(),
+                           stringify((*prev).first).c_str(), current.first.c_str(), &sd_info);
+    if (r != 0) {
+      derr << ": failed to open snapdiff, r=" << r << dendl;
+      return r;
+    }
+    while (0 < (r = ceph_readdir_snapdiff(&sd_info, &sd_entry))) {
+      if (r < 0) {
+        derr << ": failed to read directory=" << epath << dendl;
+        continue;
+      }
+
+      //New entry found
+      nname = sd_entry.dir_entry.d_name;
+      ntype = sd_entry.dir_entry.d_type;
+      // create path for the newly found entry
+      npath = entry_diff_path(epath, nname);
+      nabs_path = entry_diff_path(dir_root, npath);
+
+      struct ceph_statx cstx;
+      r = ceph_statx(sd_info.cmount, nabs_path.c_str(), &cstx,
+                     CEPH_STATX_MODE | CEPH_STATX_UID | CEPH_STATX_GID |
+                     CEPH_STATX_SIZE | CEPH_STATX_ATIME | CEPH_STATX_MTIME,
+                     AT_STATX_DONT_SYNC | AT_SYMLINK_NOFOLLOW);
+
+      if (ntype == 4) { // is a directory
```

Comment:
not like this please - use the macro defn.

Reply:
Is there an already existing one ? `O_DIRECTORY` is something else.

Reply:
macro is not appropriate here. We need an `is_directory` function in `struct ceph_snapdiff_entry_t`. But libcephfs.h doesn't allow that.

```
ceph/src/include/cephfs/libcephfs.h:119:23: error: expected ‘:’, ‘,’, ‘;’, ‘}’ or ‘__attribute__’ before ‘{’ token
  119 |   bool is_directory() {
      |                       ^
```
I can't forward declare `ceph_snapdiff_entry_t`, as I wanted to use values (not pointers) in `SyncDiffEntry`.

Reply:
You are doing a statx anyway with the mode flag - you can use S_IFDIR, isn't it?

Reply:
It's possible. But the point is, we got the entry name from `ceph_readdir_snapdiff` output -  `sd_entry`. So it's straight forward to ask what type it is to `sd_entry`. The only failure here is unable to define C++ way of `is_directory`, which I'll continue checking. Currently a function pointer is the way libcephfs.h shows, but that's also unnecessary as a getter. Let me check more on this.

Reply:
Which means you'd need to check the return value of statx first before accessing `cstx`. That's bug in the implementation.